extensions-core/s3-extensions/src/main/java/org/apache/druid/data/input/s3/S3InputSource.java

```diff
@@ -0,0 +1,259 @@
+    }
+
+    if (this.uris.isEmpty() && this.prefixes.isEmpty()) {
+      throw new IAE("uris or prefixes must be specified");
+    }
+
+    for (final URI inputURI : this.uris) {
+      Preconditions.checkArgument("s3".equals(inputURI.getScheme()), "input uri scheme == s3 (%s)", inputURI);
+    }
+
+    for (final URI inputURI : this.prefixes) {
+      Preconditions.checkArgument("s3".equals(inputURI.getScheme()), "input uri scheme == s3 (%s)", inputURI);
+    }
+  }
+
+  @JsonProperty
+  public List<URI> getUris()
+  {
+    return uris;
+  }
+
+  @JsonProperty("prefixes")
+  public List<URI> getPrefixes()
+  {
+    return prefixes;
+  }
+
+  @Override
+  public Stream<InputSplit<URI>> createSplits(InputFormat inputFormat, @Nullable SplitHintSpec splitHintSpec)
+      throws IOException
+  {
+    if (cacheSplitUris == null) {
+      initalizeSplitUris();
+    }
+    return cacheSplitUris.stream().map(InputSplit::new);
+  }
+
+  @Override
+  public int getNumSplits(InputFormat inputFormat, @Nullable SplitHintSpec splitHintSpec) throws IOException
+  {
+    if (cacheSplitUris == null) {
+      initalizeSplitUris();
+    }
+    return cacheSplitUris.size();
+  }
+
+  @Override
+  public SplittableInputSource<URI> withSplit(InputSplit<URI> split)
+  {
+    return new S3InputSource(s3Client, ImmutableList.of(split.get()), ImmutableList.of());
+  }
+
+  @Override
+  public boolean needsFormat()
+  {
+    return true;
+  }
+
+  @Override
+  protected InputSourceReader formattableReader(
+      InputRowSchema inputRowSchema,
+      InputFormat inputFormat,
+      @Nullable File temporaryDirectory
+  ) throws IOException
+  {
+    return new InputEntityIteratingReader(
+        inputRowSchema,
+        inputFormat,
+        // formattableReader() is supposed to be called in each task that actually creates segments.
+        // The task should already have only one split in parallel indexing,
+        // while there's no need to make splits using splitHintSpec in sequential indexing.
+        createSplits(inputFormat, null).map(split -> new S3Entity(s3Client, split.get())),
+        temporaryDirectory
+    );
+  }
+
+  private void initalizeSplitUris() throws IOException
+  {
+    cacheSplitUris = !uris.isEmpty() ? uris : getUrisFromPrefix(s3Client, prefixes);
+  }
+
+  public static Collection<URI> getUrisFromPrefix(ServerSideEncryptingAmazonS3 s3Client, List<URI> prefixes)
+      throws IOException
+  {
+    final List<S3ObjectSummary> objects = new ArrayList<>();
+    for (URI uri : prefixes) {
+      final String bucket = uri.getAuthority();
+      final String prefix = S3Utils.extractS3Key(uri);
+
+      try {
+        final Iterator<S3ObjectSummary> objectSummaryIterator = S3Utils.objectSummaryIterator(
+            s3Client,
+            bucket,
+            prefix,
+            MAX_LISTING_LENGTH
+        );
+        objects.addAll(Lists.newArrayList(objectSummaryIterator));
+      }
+      catch (AmazonS3Exception outerException) {
```

Comment:
Unit test coverage is missing for the exception handling here and may be worth adding to verify the retry logic.

Reply:
added tests for this and more